packages/react-router/src/router.ts

```diff
@@ -1797,91 +1778,146 @@ export class Router<
 
                 err.routerCode = routerCode
                 firstBadMatchIndex = firstBadMatchIndex ?? index
-                handleRedirectAndNotFound(match, err)
+                handleRedirectAndNotFound(getMatch(), err)
 
                 try {
                   route.options.onError?.(err)
                 } catch (errorHandlerErr) {
                   err = errorHandlerErr
-                  handleRedirectAndNotFound(match, err)
+                  handleRedirectAndNotFound(getMatch(), err)
                 }
 
-                matches[index] = match = updateMatch(match.id, () => ({
-                  ...match,
+                updateMatch(matchId, (prev) => ({
+                  ...prev,
                   error: err,
                   status: 'error',
                   updatedAt: Date.now(),
                   abortController: new AbortController(),
                 }))
               }
 
-              if (match.paramsError) {
-                handleSerialError(match.paramsError, 'PARSE_PARAMS')
-              }
+              const runBeforeLoad = async () => {
+                if (getMatch().beforeLoadPromise) {
+                  return await getMatch().beforeLoadPromise!
+                }
 
-              if (match.searchError) {
-                handleSerialError(match.searchError, 'VALIDATE_SEARCH')
-              }
+                const beforeLoadPromise =
+                  createControlledPromise<Record<string, any>>()
 
-              try {
-                const parentContext =
-                  parentMatch?.context ?? this.options.context ?? {}
+                const previousResolve = loadPromise.resolve
+                loadPromise = createControlledPromise<void>(previousResolve)
 
-                // Make sure the match has parent context set before going further
-                matches[index] = match = {
-                  ...match,
+                const parentContext = getParentContext()
+
+                updateMatch(matchId, (prev) => ({
+                  ...prev,
+                  isFetching: 'beforeLoad',
+                  beforeLoadPromise,
+                  loadPromise,
                   routeContext: replaceEqualDeep(
-                    match.routeContext,
+                    prev.routeContext,
                     parentContext,
                   ),
-                  context: replaceEqualDeep(match.context, parentContext),
+                  context: replaceEqualDeep(prev.context, parentContext),
```

Comment:
@tannerlinsley what is the difference between `context` and `routeContext` here? As in, what's purpose of having both?
I never fully understood the reasoning for both existing.

Reply:
I thought one is the current route context (`routeContext`) where the other is merged from the parent. At least, this is how the types work